### ISSUE TITLE
[Bug Report] Unusual L0 Behavior with Gemma-2b

### ISSUE THREAD

**Issue by muyo8692** [jbloomAus/SAELens#382]
If you are submitting a bug report, please fill in the following details and use the tag [bug].

**Describe the bug**
When training SAE on Gemma-2b layer 12, the L0 values remain unusually high (>1500) and are largely unresponsive to L0 lambda parameter adjustments. This behavior persists across various hyperparameter configurations:

- L0 lambda variations (50, 100, 500) show minimal impact on L0 values
- MSE losses remain around 110 at 600k steps (≈2.5B tokens)
- Even with significantly reduced learning rate (7e-8), L0 values still exceed 1500 after 2.5B tokens
- Bandwidth and initial threshold adjustments showed limited effect on controlling L0

**Code example**
Training config as follow, simply sweep l1_coefficient, I tried 1e-3, 1, 5, 5k, 500k:
50, 500 and 50k is also tried but I accidentally delete these runs log
```python
from sae_lens import LanguageModelSAERunnerConfig, SAETrainingRunner


cfg = LanguageModelSAERunnerConfig(
    architecture="jumprelu",
    model_name="gemma-2-2b",
    model_class_name="HookedTransformer",
    hook_name="blocks.12.hook_resid_post",
    hook_eval="NOT_IN_USE",
    hook_layer=12,
    hook_head_index=None,
    dataset_path="monology/pile-uncopyrighted",
    dataset_trust_remote_code=True,
    streaming=True,
    is_dataset_tokenized=False,
    context_size=1024,
    use_cached_activations=False,
    cached_activations_path=None,
    d_in=2304,
    d_sae=16384, # gemma scope has a width of 16.4k
    b_dec_init_method="zeros", # initialized to all zeros according to Anthropic
    activation_fn="relu",
    normalize_sae_decoder=True, # throughout training, we restrict the columns of W_dec to have unit norm by renormalizing after every update
    noise_scale=0.0,
    from_pretrained_path=None,
    apply_b_dec_to_input=True, # During training, we parameterise the SAE using a pre-encoder bias Bricken et al. (2023), subtracting b_dec from activations before the encoder
    decoder_orthogonal_init=False,
    decoder_heuristic_init=False, # We initialize W_dec using He-uniform initialization and rescale each latent vector to be unit norm
    init_encoder_as_decoder_transpose=True, # W_enc is initalized as the transpose of W_dec, but they are not tied afterwards
    n_batches_in_buffer=64,
    training_tokens=4000000000,
    finetuning_tokens=0,
    store_batch_size_prompts=64,
    train_batch_size_tokens=4096, # a batch size of 4,096
    normalize_activations="expected_average_only_in", # During training, activation vectors are normalized by a fixed scalar
    device="cuda",
    act_store_device="cpu",
    seed=42,
    dtype="torch.float32",
    prepend_bos=True,
    jumprelu_init_threshold=0.001,
    jumprelu_bandwidth=0.001,
    autocast=False,
    autocast_lm=False,
    compile_llm=False,
    llm_compilation_mode="default",
    compile_sae=False,
    sae_compilation_mode=None,
    adam_beta1=0,
    adam_beta2=0.999,
    mse_loss_normalization=None,
    l1_coefficient=5, # SAELens must not normalize losses in the same way as most other libraries, since a typical L1 lambda is also around 1-10 or so usually
    lp_norm=1.0, # not used for jumprelu
    scale_sparsity_penalty_by_decoder_norm=False,
    l1_warm_up_steps=10000,
    lr=7e-08,
    lr_scheduler_name="cosineannealing",
    lr_warm_up_steps=1000,
    lr_end=7e-09,
    lr_decay_steps=0,
    finetuning_method=None,
    use_ghost_grads=False,
    feature_sampling_window=2000,
    dead_feature_window=1000,
    dead_feature_threshold=1e-08,
    n_eval_batches=10,
    eval_batch_size_prompts=1,
    log_to_wandb=True,
    log_activations_store_to_wandb=False,
    log_optimizer_state_to_wandb=False,
    wandb_project="gemma_2_2b_reproduce",
    wandb_id=None,
    run_name="pile-sae-gemma-2-2b-jumprelu-lambda-5-lr-7e-8",
    wandb_entity=None,
    wandb_log_frequency=300,
    eval_every_n_wandb_logs=10,
    resume=False,
    n_checkpoints=0,
    checkpoint_path="/home/pj24001974/ku40003408/exp_env/jpn_sae/trained_saes",
    verbose=True,
    model_kwargs={},
    model_from_pretrained_kwargs={},
)

sparse_autoencoder = SAETrainingRunner(cfg).run()
```

**System Info**
Describe the characteristic of your environment:
All library was installed using [uv](https://github.com/astral-sh/uv), pyproject.toml is shown as below:
Note: I also tried former version of sae-lens (4.3.4, 4.3.5)
```
requires-python = ">=3.12"
dependencies = [
    "circuitsvis>=1.41.0",
    "gpustat>=1.1.1",
    "huggingface-hub>=0.26.2",
    "notebook>=7.2.2",
    "sae-lens==4.4.1",
    "torch>=2.5.1",
    "transformer-lens>=2.8.1",
]
```

**Run log**
<img width="1064" alt="Screenshot 2024-11-22 at 10 11 58" src="https://github.com/user-attachments/assets/ef45bbe5-2278-4487-91c5-47ccc6558c8f">
<img width="1066" alt="Screenshot 2024-11-22 at 10 12 50" src="https://github.com/user-attachments/assets/0a645149-8580-4358-af09-86fe5969eef8">

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/jbloomAus/SAELens/issues) in the repo (**required**)

**Comment by cwyoon-99** [comment]
@muyo8692  In this moment, Have you found any solutions related to this issue? 
I have a similar problem that L0 does not converge into reasonable level (<= 150).

**Comment by Hetens** [comment]
@muyo8692 @cwyoon-99 I think I found the cause of most of this. It's the same root cause as #494, fixed in #718.

The JumpReLU threshold was stored as `log_threshold`, with `threshold = exp(log_threshold)`. Adam normalizes by gradient magnitude, so it moves a parameter by at most ~`lr` per step no matter how large the gradient is. In log space, a step of `lr` moves the *effective* threshold by only ~`threshold * lr`,  about `1e-7`/step at the `0.01` default init. The threshold is effectively frozen, and since the L0 penalty's only gradient path is the threshold, L0 can't come down.

That explains three of the things you saw:

- **L0 lambda had minimal impact.** Adam divides the gradient by its own magnitude, so a constant multiplier on the loss cancels out entirely.
- **Lowering the lr to 7e-8 didn't help, and shouldn't.** Reachable threshold is roughly `lr * n_steps`, so a smaller lr freezes it further. At `7e-8` over 600k steps the threshold can travel about `0.04` in total.
- **Bandwidth adjustments had limited effect.** I measured the fraction of pre-activations inside the STE gradient window on `main` and got ~3.2% on average.The gradient was reaching the threshold fine, it just couldn't translate into movement.

Controlled runs varying only the parameterization (gpt2 `blocks.5.hook_resid_post`, 2000 steps, `d_sae=3072`): threshold `0.0157` / L0 3019 with `exp(log_threshold)`, vs threshold `0.6197` / L0 1021 with a direct threshold. Runs: https://wandb.ai/hetansh-shah95-umass-amherst/saelens-494-jumprelu-threshold

This removes the block but doesn't by itself get you to a low L0. `lr * n_steps` still bounds how far the threshold can go, and the coefficient remains a weak knob even after the fix. I think I'm gonna open a separate issue for that part.
